msal/src/main/java/com/microsoft/identity/client/SingleAccountPublicClientApplication.java

```diff
@@ -93,39 +224,168 @@ public void onTaskCompleted(List<ICacheRecord> cacheRecords) {
+    protected ILocalAuthenticationCallback getLocalAuthenticationCallback(final AuthenticationCallback authenticationCallback) {
+
+        return new ILocalAuthenticationCallback() {
+
+            @Override
+            public void onSuccess(ILocalAuthenticationResult localAuthenticationResult) {
+
+                //Get Local Authentication Result then check if the current account is set or not
+
+                MultiTenantAccount newAccount =  getAccountFromICacheRecordList(localAuthenticationResult.getCacheRecordWithTenantProfileData());
+
+                if(didExistingCurrentAccountChange(newAccount)){
+                    //Throw on Error with UserMismatchException
+                    authenticationCallback.onError(new MsalClientException(MsalClientException.CURRENT_ACCOUNT_MISMATCH));
+                    return;
+                }else{
+                    persistCurrentAccount(localAuthenticationResult.getCacheRecordWithTenantProfileData());
+                }
+
+                IAuthenticationResult authenticationResult = AuthenticationResultAdapter.adapt(localAuthenticationResult);
+                authenticationCallback.onSuccess(authenticationResult);
+            }
+
+            @Override
+            public void onError(BaseException exception) {
+                MsalException msalException = MsalExceptionAdapter.msalExceptionFromBaseException(exception);
+                authenticationCallback.onError(msalException);
+            }
+
+            @Override
+            public void onCancel() {
+                authenticationCallback.onCancel();
+            }
+        };
+    }
+
+    private boolean didExistingCurrentAccountChange(MultiTenantAccount newAccount){
+        MultiTenantAccount persistedAccount = getPersistedCurrentAccount();
+        return (persistedAccount != null || newAccount.getId() != persistedAccount.getId()) ? true : false;
+
+    }
+
+    @Override
+    public void signOut(@NonNull final TaskCompletedCallbackWithError<Boolean, Exception> callback){
+        final String methodName = ":signOut";
         final PublicClientApplicationConfiguration configuration = getConfiguration();
 
-        new BrokerMsalController().removeAccountFromSharedDevice(
-                configuration,
-                new TaskCompletedCallbackWithError<Boolean, Exception>() {
-                    @Override
-                    public void onTaskCompleted(Boolean success) {
-                        if (success) {
-                            persistCurrentAccount(null);
-                        }
 
-                        callback.onTaskCompleted(success);
-                    }
+        try{
+            if (MSALControllerFactory.brokerEligible(
+                    configuration.getAppContext(),
+                    configuration.getDefaultAuthority(),
+                    configuration) && mIsSharedDevice) {
```

Comment:
BrokerEligible() is already invoked in BrokerMsalController. removeAccountFromSharedDevice()

Reply:
Agreed.